x/claim/keeper/claim.go

```diff
@@ -0,0 +1,551 @@
+		Mul(claimRecord.Weight).
+		Quo(totalWeight).RoundInt()
+	claimableCoins := sdk.NewCoins(sdk.NewCoin(airdrop.ClaimDenom, claimableAmount))
+
+	elapsedAirdropTime := ctx.BlockTime().Sub(airdrop.AirdropStartTime)
+	// The entire airdrop has completed
+	if elapsedAirdropTime > airdrop.AirdropDuration {
+		return sdk.Coins{}, nil
+	}
+	return claimableCoins, nil
+}
+
+// GetClaimable returns claimable amount for a specific action done by an address
+func (k Keeper) GetUserTotalClaimable(ctx sdk.Context, addr sdk.AccAddress, airdropIdentifier string) (sdk.Coins, error) {
+	claimRecord, err := k.GetClaimRecord(ctx, addr, airdropIdentifier)
+	if err != nil {
+		return sdk.Coins{}, err
+	}
+	if claimRecord.Address == "" {
+		return sdk.Coins{}, nil
+	}
+
+	totalClaimable := sdk.Coins{}
+
+	for action := range types.Action_name {
+		claimableForAction, err := k.GetClaimableAmountForAction(ctx, addr, types.Action(action), airdropIdentifier)
+		if err != nil {
+			return sdk.Coins{}, err
+		}
+		if !claimableForAction.Empty() {
+			totalClaimable = totalClaimable.Add(claimableForAction...)
+		}
+	}
+	return totalClaimable, nil
+}
+
+// Get airdrop identifier corresponding to the user address
+func (k Keeper) GetAirdropIdentifiersForUser(ctx sdk.Context, addr sdk.AccAddress) []string {
+	store := ctx.KVStore(k.storeKey)
+	params, err := k.GetParams(ctx)
+	identifiers := []string{}
+	if err != nil {
+		return identifiers
+	}
+
+	for _, airdrop := range params.Airdrops {
+		prefixStore := prefix.NewStore(store, append([]byte(types.ClaimRecordsStorePrefix), []byte(airdrop.AirdropIdentifier)...))
+		if prefixStore.Has(addr) {
+			identifiers = append(identifiers, airdrop.AirdropIdentifier)
+		}
+	}
+	return identifiers
+}
+
+// ClaimCoins remove claimable amount entry and transfer it to user's account
+func (k Keeper) ClaimCoinsForAction(ctx sdk.Context, addr sdk.AccAddress, action types.Action, airdropIdentifier string) (sdk.Coins, error) {
+	if airdropIdentifier == "" {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "invalid airdrop identifier: ClaimCoinsForAction")
+	}
+
+	claimableAmount, err := k.GetClaimableAmountForAction(ctx, addr, action, airdropIdentifier)
+	if err != nil {
+		return claimableAmount, err
+	}
+
+	if claimableAmount.Empty() {
+		return claimableAmount, nil
+	}
+
+	claimRecord, err := k.GetClaimRecord(ctx, addr, airdropIdentifier)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the account is a default vesting account, don't grant new vestings
+	acc := k.accountKeeper.GetAccount(ctx, addr)
+	_, isDefaultVestingAccount := acc.(*authvestingtypes.BaseVestingAccount)
+	if isDefaultVestingAccount {
+		return nil, err
+	}
+
+	acc = k.accountKeeper.GetAccount(ctx, addr)
+	strideVestingAcc, isStrideVestingAccount := acc.(*vestingtypes.StridePeriodicVestingAccount)
+	// Check if vesting tokens already exist for this account.
+	if !isStrideVestingAccount {
+		// Convert user account into stride veting account.
```

Comment:
What happens if there are existing tokens in the account? Will they be forced to vest too? (I don't think they should, just double checking)

Reply:
No. there are no impacts on the existing tokens.